src/wallet/asyncrpcoperation_shieldcoinbase.cpp

```diff
@@ -157,303 +107,136 @@ void AsyncRPCOperation_shieldcoinbase::main() {
-    }
+    return Remaining(utxoCounter, numUtxos, remainingValue, shieldingValue);
+}
 
-    KeyIO keyIO(Params());
-
-    // !!! Payment disclosure START
-    size_t js_index = tx_.vJoinSplit.size() - 1;
-    uint256 placeholder;
-    for (int i = 0; i < ZC_NUM_JS_OUTPUTS; i++) {
-        uint8_t mapped_index = outputMap[i];
-        // placeholder for txid will be filled in later when tx has been finalized and signed.
-        PaymentDisclosureKey pdKey = {placeholder, js_index, mapped_index};
-        JSOutput output = outputs[mapped_index];
-        libzcash::SproutPaymentAddress zaddr = output.addr;  // randomized output
-        PaymentDisclosureInfo pdInfo = {PAYMENT_DISCLOSURE_VERSION_EXPERIMENTAL, esk, joinSplitPrivKey_, zaddr};
-        paymentDisclosureData_.push_back(PaymentDisclosureKeyInfo(pdKey, pdInfo));
-
-        LogPrint("paymentdisclosure", "%s: Payment Disclosure: js=%d, n=%d, zaddr=%s\n", getId(), js_index, int(mapped_index), keyIO.EncodePaymentAddress(zaddr));
-    }
-    // !!! Payment disclosure END
-
-    UniValue obj(UniValue::VOBJ);
-    obj.pushKV("encryptednote1", encryptedNote1);
-    obj.pushKV("encryptednote2", encryptedNote2);
-    obj.pushKV("rawtxn", HexStr(ss.begin(), ss.end()));
-    obj.pushKV("inputmap", arrInputMap);
-    obj.pushKV("outputmap", arrOutputMap);
-    return obj;
+uint256 AsyncRPCOperation_shieldcoinbase::main_impl() {
+    uint256 txid;
+
+    const auto& spendable = effects_->GetSpendable();
+    const auto& payments = effects_->GetPayments();
+    spendable.LogInputs(getId());
+
+    LogPrint("zrpcunsafe", "%s: spending %s to send %s with fee %s\n", getId(),
+        FormatMoney(payments.Total()),
+        FormatMoney(spendable.Total()),
+        FormatMoney(effects_->GetFee()));
+    LogPrint("zrpc", "%s: total transparent input: %s (to choose from)\n", getId(),
+        FormatMoney(spendable.GetTransparentTotal()));
+    LogPrint("zrpcunsafe", "%s: total shielded input: %s (to choose from)\n", getId(),
+        FormatMoney(spendable.GetSaplingTotal() + spendable.GetOrchardTotal()));
+    LogPrint("zrpc", "%s: total transparent output: %s\n", getId(),
+        FormatMoney(payments.GetTransparentTotal()));
+    LogPrint("zrpcunsafe", "%s: total shielded Sapling output: %s\n", getId(),
+        FormatMoney(payments.GetSaplingTotal()));
+    LogPrint("zrpcunsafe", "%s: total shielded Orchard output: %s\n", getId(),
+        FormatMoney(payments.GetOrchardTotal()));
+    LogPrint("zrpc", "%s: fee: %s\n", getId(), FormatMoney(effects_->GetFee()));
+
+    auto buildResult = effects_->ApproveAndBuild(
+            Params().GetConsensus(),
+            *pwalletMain,
+            chainActive,
+            strategy_);
+
+    auto tx = buildResult.GetTxOrThrow();
+
+    UniValue sendResult = SendTransaction(tx, payments.GetResolvedPayments(), std::nullopt, testmode);
+    set_result(sendResult);
+
+    txid = tx.GetHash();
+
+    effects_->UnlockSpendable(*pwalletMain);
```

Comment:
Where were these locked? And if somewhere earlier inside `WalletTxBuilder`, why does this not also happen inside `WalletTxBuilder`?

Reply:
They are locked immediately after creating the `TransactionEffects`. I agree the asymmetry isn’t ideal, but we need to hold onto the lock until after we’ve sent the transaction, plus we need to handle the failure case where we never call `SendTransaction` (which this code is missing – fixing that!), so we can’t unlock in there (or at least not only in there, and so having all the unlocking in the same place is helpful).

In future, we’d like to hold the locks until after the expiry height, but this implementation matches the current semantics and avoids a bigger change for now.

Reply:
I could move `LockSpendable` out of `WalletTxBuilder` to here – I’d like to get them both _into_ WalletTxBuilder, but that isn’t feasible at the moment.

Reply:
Even moving them out, they’re in separate methods, because locking happens after preparation, and unlocking after approve/build/send. So, while it’s necessary to have the locking already, we want them to be in the right places so they just work when we switch to 2-stage tx.

Tangentially (and likely not for this PR) … z_shieldcoinbase already does preparation synchronously, since it returns some extra data that we get via preparation. Should _all_ of the async ops have sync preparation? Both for consistency and for 2-stage prep.

Reply:
I don't think that synchronous prep makes much difference here; the caller still has to check the async result for failures irrespective whether prep succeeded or failed.

Reply:
When you say “here”, what do you mean?
- in z_shieldcoinbase? In this case, we need to do it sync, just because z_shieldcoinbase return ancillary info that’s only available once we prepare.
- in general, before 2-stage? I agree it doesn’t make a lot of difference, which makes me inclined to do sync prep (after this release) simply because with the consistency, we can maybe move more logic to `wallet_tx_builder` `async…_common`, etc. rather than duplicating between calls; and also because that’s roughly how 2-stage tx will work (right?)